File: src/handlers/tokens.js
import { json } from '../utils/response';
import { executeGraphQLQuery } from '../utils/graphql';

/**
 * Get all tokens with their metadata with pagination
 * @param {Request} request - The incoming request
 * @returns {Promise<Response>} - The response with token data
 */
export async function getAllTokens(request) {
  const url = new URL(request.url);
  const offset = parseInt(url.searchParams.get('offset') || '0', 10);
  const limit = parseInt(url.searchParams.get('limit') || '10', 10);
  
  // Cap the limit to prevent excessive queries
  const safeLimit = Math.min(limit, 20);
  
  try {
    // First query - get all contracts with pagination
    const contractListQuery = `
      query TokenContracts {
        allContracts(
          first: ${safeLimit}
          offset: ${offset}
          filter: {xsc0001: {equalTo: true}}
          orderBy: CREATED_DESC
        ) {
          totalCount
          nodes { 
            name 
            created 
          }
        }
      }
    `;
    
    const contractsData = await executeGraphQLQuery(contractListQuery);
    
    const nodes = contractsData?.data?.allContracts?.nodes || [];
    const totalCount = contractsData?.data?.allContracts?.totalCount || 0;
    
    if (!nodes.length) {
      return json({
        tokens: [],
        pagination: {
          offset,
          limit: safeLimit,
          total: totalCount
        }
      });
    }
    
    // Build the list of metadata keys we need
    const metaKeys = [];
    nodes.forEach(({ name }) => {
      metaKeys.push(`${name}.metadata:token_name`);
      metaKeys.push(`${name}.metadata:token_symbol`);
      metaKeys.push(`${name}.metadata:token_logo_url`);
      metaKeys.push(`${name}.metadata:token_website`);
      metaKeys.push(`${name}.metadata:total_supply`);
      metaKeys.push(`${name}.metadata:operator`);
    });
    
    // Second query - pull the metadata in one call
    const metaQuery = `
      query TokenMeta {
        allStates(filter:{ key:{ in:[${metaKeys.map(k => `"${k}"`).join(',')}] } }) {
          edges { node { key value } }
        }
      }
    `;
    
    const metaResp = await executeGraphQLQuery(metaQuery);
    const metaEdges = metaResp?.data?.allStates?.edges || [];
    
    // Build a lookup: { con_usdc: { token_name:'USDC', ... } }
    const metaMap = {};
    metaEdges.forEach(({ node }) => {
      const [contractDotMeta, field] = node.key.split(":");
      const contract = contractDotMeta.replace(".metadata", "");
      if (!metaMap[contract]) metaMap[contract] = {};
      metaMap[contract][field] = node.value;
    });
    
    // Final combine & format
    const tokens = nodes.map(c => {
      const m = metaMap[c.name] || {};
      const display = m.token_name
        ? `${m.token_name}${m.token_symbol ? " (" + m.token_symbol + ")" : ""}`
        : c.name;
      
      return {
        contractName: c.name,
        token_name: m.token_name || null,
        token_symbol: m.token_symbol || null,
        token_logo_url: m.token_logo_url || null,
        token_website: m.token_website || null,
        total_supply: m.total_supply ? parseFloat(m.total_supply) : null,
        operator: m.operator || null,
        display,
        created_at: c.created
      };
    });
    
    return json({
      tokens,
      pagination: {
        offset,
        limit: safeLimit,
        total: totalCount,
        next: offset + safeLimit < totalCount ? offset + safeLimit : null,
        previous: offset > 0 ? Math.max(0, offset - safeLimit) : null
      }
    });
  } catch (error) {
    console.error('Error fetching tokens:', error);
    return new Response(JSON.stringify({ error: 'Failed to fetch tokens' }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' }
    });
  }
}
export async function getTokenByName(request, { contractName }) {
  try {
    const contractNames = decodeURIComponent(contractName).split(',').map(name => name.trim()).filter(Boolean);

    console.log(`Getting token data for: ${contractNames.join(', ')}`);

    // Generate dynamic GraphQL filters
    const nameFilters = contractNames.map(name => `{ equalTo: "${name}" }`).join(", ");
    const stateKeys = contractNames.flatMap(name => [
      `${name}.metadata:token_name`,
      `${name}.metadata:token_symbol`,
      `${name}.metadata:token_logo_url`,
      `${name}.metadata:token_website`,
      `${name}.metadata:total_supply`,
      `${name}.metadata:operator`,
    ]);

        const query = `
      query GetTokenData($names: [String!], $stateKeys: [String!], $firstContracts: Int!, $firstStates: Int!) {
        allContracts(
          filter: {
            xsc0001: {equalTo: true},
            name: {in: $names}
          },
          first: $firstContracts
        ) {
          nodes {
            name
            created
         }
        }
        allStates(
          filter: { key: { in: $stateKeys } },
          first: $firstStates
        ) {
          edges {
            node {
              key
              value
            }
          }
        }
      }
    `;

    console.log(`Executing query: ${query}`);
    const data = await executeGraphQLQuery(query, {
      names: contractNames,
      stateKeys,
      firstContracts: contractNames.length,    // enough to cover all requested names
      firstStates:    stateKeys.length         // enough to cover every metadata key
    });
    console.log(`Query result: ${JSON.stringify(data)}`);

    const contracts = data?.data?.allContracts?.nodes || [];
    const metaEdges = data?.data?.allStates?.edges || [];

    const results = contractNames.map(name => {
      const contract = contracts.find(c => c.name === name);
      if (!contract) {
        return {
          contractName: name,
          error: 'Token contract not found',
          message: 'The specified contract does not exist or is not a token (XSC-0001 standard)'
        };
      }

      const metadata = {};
      metaEdges.forEach(({ node }) => {
        if (node.key.startsWith(`${name}.metadata:`)) {
          const field = node.key.split(":")[1];
          metadata[field] = node.value;
        }
      });

      return {
        contractName: contract.name,
        token_name: metadata.token_name || null,
        token_symbol: metadata.token_symbol || null,
        token_logo_url: metadata.token_logo_url || null,
        token_website: metadata.token_website || null,
        total_supply: metadata.total_supply ? parseFloat(metadata.total_supply) : null,
        operator: metadata.operator || null,
        display: metadata.token_name
          ? `${metadata.token_name}${metadata.token_symbol ? " (" + metadata.token_symbol + ")" : ""}`
          : contract.name,
        created_at: contract.created
      };
    });

    return json(results.length === 1 ? results[0] : results);

  } catch (error) {
    console.error('Error fetching token metadata:', error);
    return new Response(JSON.stringify({ 
      error: 'Failed to fetch token metadata',
      message: error.message || 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' }
    });
  }
}

File: app/handlers/tokens.py
"""
Handlers for token-related endpoints
"""
from typing import Dict, List, Optional, Union, Any
from fastapi import Request, HTTPException

from app.utils.graphql import execute_graphql_query
from app.utils.response import json_response


async def get_all_tokens(request: Request):
    """
    Get all tokens with their metadata with pagination
    
    Args:
        request: The incoming request
        
    Returns:
        JSON response with token data
    """
    url = request.url
    offset = int(request.query_params.get('offset', '0'))
    limit = int(request.query_params.get('limit', '10'))
    
    # Cap the limit to prevent excessive queries
    safe_limit = min(limit, 20)
    
    try:
        # First query - get all contracts with pagination
        contract_list_query = f"""
            query TokenContracts {{
                allContracts(
                    first: {safe_limit}
                    offset: {offset}
                    filter: {{xsc0001: {{equalTo: true}}}}
                    orderBy: CREATED_DESC
                ) {{
                    totalCount
                    nodes {{ 
                        name 
                        created 
                    }}
                }}
            }}
        """
        
        contracts_data = await execute_graphql_query(contract_list_query)
        
        nodes = contracts_data.get('data', {}).get('allContracts', {}).get('nodes', [])
        total_count = contracts_data.get('data', {}).get('allContracts', {}).get('totalCount', 0)
        
        if not nodes:
            return json_response({
                "tokens": [],
                "pagination": {
                    "offset": offset,
                    "limit": safe_limit,
                    "total": total_count
                }
            })
        
        # Build the list of metadata keys we need
        meta_keys = []
        for node in nodes:
            name = node.get('name')
            meta_keys.append(f"{name}.metadata:token_name")
            meta_keys.append(f"{name}.metadata:token_symbol")
            meta_keys.append(f"{name}.metadata:token_logo_url")
            meta_keys.append(f"{name}.metadata:token_website")
            meta_keys.append(f"{name}.metadata:total_supply")
            meta_keys.append(f"{name}.metadata:operator")
        
        # Second query - pull the metadata in one call
        meta_query = f"""
            query TokenMeta {{
                allStates(filter:{{ key:{{ in:[{','.join([f'"{k}"' for k in meta_keys])}] }} }}) {{
                    edges {{ node {{ key value }} }}
                }}
            }}
        """
        
        meta_resp = await execute_graphql_query(meta_query)
        meta_edges = meta_resp.get('data', {}).get('allStates', {}).get('edges', [])
        
        # Build a lookup: { con_usdc: { token_name:'USDC', ... } }
        meta_map = {}
        for edge in meta_edges:
            node = edge.get('node', {})
            key = node.get('key', '')
            value = node.get('value', '')
            
            contract_dot_meta, field = key.split(":")
            contract = contract_dot_meta.replace(".metadata", "")
            
            if contract not in meta_map:
                meta_map[contract] = {}
            
            meta_map[contract][field] = value
        
        # Final combine & format
        tokens = []
        for c in nodes:
            name = c.get('name')
            m = meta_map.get(name, {})
            
            token_name = m.get('token_name')
            token_symbol = m.get('token_symbol')
            
            display = name
            if token_name:
                display = token_name
                if token_symbol:
                    display += f" ({token_symbol})"
            
            tokens.append({
                "contractName": name,
                "token_name": token_name,
                "token_symbol": token_symbol,
                "token_logo_url": m.get('token_logo_url'),
                "token_website": m.get('token_website'),
                "total_supply": float(m.get('total_supply')) if m.get('total_supply') else None,
                "operator": m.get('operator'),
                "display": display,
                "created_at": c.get('created')
            })
        
        return json_response({
            "tokens": tokens,
            "pagination": {
                "offset": offset,
                "limit": safe_limit,
                "total": total_count,
                "next": offset + safe_limit if offset + safe_limit < total_count else None,
                "previous": max(0, offset - safe_limit) if offset > 0 else None
            }
        })
    except Exception as e:
        print(f"Error fetching tokens: {e}")
        return json_response(
            {"error": "Failed to fetch tokens", "message": str(e)},
            status_code=500
        )


async def get_token_by_name(request: Request, contract_name: str):
    """
    Get token by name
    
    Args:
        request: The incoming request
        contract_name: The contract name
        
    Returns:
        JSON response with token data
    """
    try:
        contract_names = contract_name.split(',')
        contract_names = [name.strip() for name in contract_names if name.strip()]
        
        print(f"Getting token data for: {', '.join(contract_names)}")
        
        # Generate dynamic GraphQL filters
        state_keys = []
        for name in contract_names:
            state_keys.extend([
                f"{name}.metadata:token_name",
                f"{name}.metadata:token_symbol",
                f"{name}.metadata:token_logo_url",
                f"{name}.metadata:token_website",
                f"{name}.metadata:total_supply",
                f"{name}.metadata:operator",
            ])
        
        query = """
            query GetTokenData($names: [String!], $stateKeys: [String!], $firstContracts: Int!, $firstStates: Int!) {
                allContracts(
                    filter: {
                        xsc0001: {equalTo: true},
                        name: {in: $names}
                    },
                    first: $firstContracts
                ) {
                    nodes {
                        name
                        created
                    }
                }
                allStates(
                    filter: { key: { in: $stateKeys } },
                    first: $firstStates
                ) {
                    edges {
                        node {
                            key
                            value
                        }
                    }
                }
            }
        """
        
        variables = {
            "names": contract_names,
            "stateKeys": state_keys,
            "firstContracts": len(contract_names),  # enough to cover all requested names
            "firstStates": len(state_keys)          # enough to cover every metadata key
        }
        
        data = await execute_graphql_query(query, variables)
        
        contracts = data.get('data', {}).get('allContracts', {}).get('nodes', [])
        meta_edges = data.get('data', {}).get('allStates', {}).get('edges', [])
        
        results = []
        for name in contract_names:
            contract = next((c for c in contracts if c.get('name') == name), None)
            
            if not contract:
                results.append({
                    "contractName": name,
                    "error": "Token contract not found",
                    "message": "The specified contract does not exist or is not a token (XSC-0001 standard)"
                })
                continue
            
            metadata = {}
            for edge in meta_edges:
                node = edge.get('node', {})
                key = node.get('key', '')
                
                if key.startswith(f"{name}.metadata:"):
                    field = key.split(":")[1]
                    metadata[field] = node.get('value', '')
            
            token_name = metadata.get('token_name')
            token_symbol = metadata.get('token_symbol')
            
            display = contract.get('name')
            if token_name:
                display = token_name
                if token_symbol:
                    display += f" ({token_symbol})"
            
            results.append({
                "contractName": contract.get('name'),
                "token_name": token_name,
                "token_symbol": token_symbol,
                "token_logo_url": metadata.get('token_logo_url'),
                "token_website": metadata.get('token_website'),
                "total_supply": float(metadata.get('total_supply')) if metadata.get('total_supply') else None,
                "operator": metadata.get('operator'),
                "display": display,
                "created_at": contract.get('created')
            })
        
        return json_response(results[0] if len(results) == 1 else results)
    
    except Exception as e:
        print(f"Error fetching token metadata: {e}")
        return json_response(
            {
                "error": "Failed to fetch token metadata",
                "message": str(e)
            },
            status_code=500
        )